### PR TITLE
Native + Lucene Fusion index as default

### PR DESCRIPTION
**Pull request by burqen**
## Fusion index as default
Fusion index = a fusion between native and Lucene index.
Native index is built on top of GB+Tree and can only handle numbers.
Lucene index is old Lucene implementation.

Fusion index works by wrapping Native and Lucene into one logical index where all numbers are directed to Native and the rest (including all arrays and all composite keys) to Lucene.

First benchmarks show a write performance improvement of between 1.5-4x.

## Default behaviour
- All newly created indexes will be 'fusion indexes'.
- Old indexes (created with only Lucene) will not be migrated but continue as 'Lucene only' indexes. Drop and recreate to use fusion index. This is recommended for indexes on number focused properties (not composite or arrays). For indexes over Strings there is no use to move to fusion.

## Migration
Migrating from <3.3 will keep all indexes around and will work just fine as Lucene only. Drop and recreate to move to fusion index.

## Internal config to disable
An internal unsupported config option is introduced to disable fusion index as default. In this case, Lucene will be used instead. Disabling fusion index will NOT migrate already created fusion indexes. Those will have to be dropped and recreated to use the new default (Lucene).

## TL;DR
- New fusion index = native + lucene.
- Fusion is default.
- When creating new index, default is used.
- No migration is done for already existing indexes. Can be migrated to default by drop and recreate.
- Turn fusion index of with internal unsupported config option.
- Faster, Harder, Scooter.

co-author: @tinwelint 

[cl] Introduce new native index for numbers.

### CODE REVIEW

File: community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java
Comment (MishaDemianenko) [inline community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: Please fields first, public member and them private member

Comment (MishaDemianenko) [inline community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: not sure what collectors class and definition does in SchemaIndexExtensionLoader, even if it's passed as one of the parameters to instantiate extensions

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: `private List<T> lowerPrioritizedCandidates = Collections.emptyList();` then you do not need to suppress warnings

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: `<T extends Comparable>` instead `<T>`

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: no need to suppress warnings

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: ` List<T> all = (List<T>) Iterables.asList( candidates );` instead

File: community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java
Comment (MishaDemianenko) [inline community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:565]: what's the point doing force specifically before closing accessor?

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java
Comment (MishaDemianenko) [inline community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java:392]: noo, that was one of my favourite non-readable statements 😞

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:404]: surprising option

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java]: this method is not used at all now

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java]: there is no need to pass second param here since population is already passed as first one

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java]: feature toggle is moved but class is the same? why?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:82]: what is the sense in throwing and catching useless exceptions instead of updating providers to not throw exceptions if not failed?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:98]: please use `StringUtils.isNotEmpty` instead

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:416]: TODO

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java:1]: Not quite sure why this test was removed, we do not cancel sampling anymore?

**Review by MishaDemianenko — CHANGES_REQUESTED** [review]
Couple of small comments here and there

File: community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java
Comment (tinwelint) [reply community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: Well, this class is a utility used in a bunch of places in consistency checker. It loads schema index provider kernel extensions and consistency checker is not allowed to do recovery. Where should it go?

File: community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java
Comment (tinwelint) [reply community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:565]: That's the general contract and something that native indexing hooks in on to checkpoint properly. It's just that lucene happened to make do without it.

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java
Comment (tinwelint) [reply community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java:392]: Bugger, well I'd guess it wasn't the last of its kind :)

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:404]: We went for it since it's easier to manage, also in tests. Now that we've gotten all fixed and working we can investigate if it would be tricky to manage this if it were a toggle instead. Hang on.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java]: Something of a confusion move at a time we were confused about multi index populator. Forgot about it... we can move it back.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:98]: I think we want to check `null` explicitly. Empty string means something else in this case.

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:416]: Ah, to be investigated...

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java
Comment (tinwelint) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java:1]: It only tested that the in-memory test index `HashBasedIndex` sample method threw exception in a particular case. I think it was mis-written to begin with.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:82]: `getPopulationFailure` is designed to throw `IllegalStateException` if called for index that isn't in `FAILED` state, sensible methinks.

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:416]: It was left there because that's how it was previously. Now the `SchemaIndexProviderMap` is used instead. It's not strictly required for us to remove this right now, since it may require some time and testing. We've created a card for it now.

**Comment by tinwelint** [comment]
I think we need to look at `SchemaIndexHaIT` dude, it fails consisently on this branch... not very often at all otherwise. I think it's simply a kernel extension loading thingie.

File: community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java
Comment (MishaDemianenko) [reply community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: well if you will need in any other test ever RecoveryPreventingCollector I would say that SchemaIndexExtensionLoader would be the last place where i would expect to find it, even in separate class is better imho

File: community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java
Comment (MishaDemianenko) [reply community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:565]: I mean if you close it without force it should do that automatically right?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java
Comment (MishaDemianenko) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java]: The option still present in this class

**Comment by MishaDemianenko** [comment]
@tinwelint @burqen there is one small thing still please fix that otherwise its good and approved

**Review by MishaDemianenko — APPROVED** [review]

**Comment by tinwelint** [comment]
The build passed, although the Github <--> CI integration didn't quite get that bit. @burqen was there some issue with leaving open files behind (as shown in some benchmark)? If so perhaps we should hold off merge until we know more.

**Comment by burqen** [comment]
@tinwelint It's a bit unclear. There does not seem to be any leaking files but we need some more investigation before merging.

**Pull request merged**
